hazelcast/src/main/java/com/hazelcast/instance/impl/ClusterTopologyIntentTrackerImpl.java

```diff
@@ -0,0 +1,292 @@
+    /**
+     * @return next {@code ClusterTopologyIntent} according to rules applicable while there is no change
+     * in cluster size specification in managed context. (ie StatefulSetSpec.size remains the same, so
+     * user does not intend a change in cluster size).
+     */
+    private ClusterTopologyIntent nextIntentWhenClusterSpecEqual(ClusterTopologyIntent previous,
+                                                                 int readyNodesCount,
+                                                                 int currentNodesCount) {
+        if (readyNodesCount == currentClusterSpecSize && previous != ClusterTopologyIntent.CLUSTER_STABLE) {
+            return ClusterTopologyIntent.CLUSTER_STABLE;
+        } else if (previous == ClusterTopologyIntent.CLUSTER_STABLE && currentNodesCount < currentClusterSpecSize) {
+            return ClusterTopologyIntent.CLUSTER_STABLE_WITH_MISSING_MEMBERS;
+        } else {
+            return previous;
+        }
+    }
+
+    @Override
+    public ClusterTopologyIntent getClusterTopologyIntent() {
+        return clusterTopologyIntent.get();
+    }
+
+    @Override
+    public void initializeClusterTopologyIntent(ClusterTopologyIntent clusterTopologyIntent) {
+        ClusterTopologyIntent current = this.clusterTopologyIntent.get();
+        logger.info("Current node cluster topology intent is " + current);
+        // if not UNKNOWN, then it was already initialized
+        if (current == ClusterTopologyIntent.IN_MANAGED_CONTEXT_UNKNOWN) {
+            logger.info("Initializing this node's cluster topology to " + clusterTopologyIntent);
+            this.clusterTopologyIntent.set(clusterTopologyIntent);
+        }
+    }
+
+    @Override
+    public void shutdownWithIntent(ClusterTopologyIntent shutdownIntent) {
+        // consider the detected shutdown intent before triggering node shutdown
+        if (shutdownIntent == ClusterTopologyIntent.CLUSTER_STABLE
+                || shutdownIntent == ClusterTopologyIntent.CLUSTER_STABLE_WITH_MISSING_MEMBERS) {
+            try {
+                long timeoutNanos = node.getProperties().getNanos(ClusterProperty.CLUSTER_SHUTDOWN_TIMEOUT_SECONDS);
+                while (!getPartitionService().isPartitionTableHealthy()
+                        && timeoutNanos > 0) {
+                    sleep(TimeUnit.SECONDS.toMillis(1));
+                    timeoutNanos -= TimeUnit.SECONDS.toNanos(1);
+                }
+                changeClusterState(clusterStateForMissingMembers);
+            } catch (Throwable t) {
+                // let shutdown proceed even though we failed to switch to FROZEN state
+                logger.warning("Could not switch to transient FROZEN state while cluster"
+                        + "shutdown intent was " + shutdownIntent, t);
+            }
+        } else if (shutdownIntent == ClusterTopologyIntent.CLUSTER_SHUTDOWN
+                || shutdownIntent == ClusterTopologyIntent.CLUSTER_SHUTDOWN_WITH_MISSING_MEMBERS) {
```

Comment:
can't we also enable this auto-shutdown-all behavior when there is no persistence? I feel like this will be a good enhancement for also big memory clusters, shutdowns will be faster.